crates/edr_solidity/src/contracts_identifier/radix_tree.rs

```diff
@@ -0,0 +1,540 @@
+
+            // create a new node for the word
+            let new_node = RadixNode::new(
+                word.slice(prefix_length..),
+                true,
+                middle_node.bytes_matched_before + middle_node.content.len(),
+            );
+            middle_node
+                .child_nodes
+                .insert(word[prefix_length], new_node);
+
+            // set the middle_node as current_node's child
+            self.child_nodes.insert(b, middle_node);
+        } else {
+            let bytes_matched_before = self.bytes_matched_before + self.content.len();
+
+            let node = RadixNode::new(word, true, bytes_matched_before);
+
+            self.child_nodes.insert(b, node);
+        }
+    }
+
+    /**
+     * Returns a tuple containing:
+     * - a boolean indicating if the word was matched exactly
+     * - the number of bytes matched
+     * - the node that matched the word
+     * If the word is not matched exactly, the node will be the one that
+     * matched the longest prefix.
+     */
+    pub fn get_max_match(&self, word: &[u8]) -> (bool, usize, &RadixNode) {
+        let prefix_length = get_shared_prefix_length(word, &self.content);
+
+        let matched = prefix_length + self.bytes_matched_before;
+
+        let entire_word_matched = prefix_length == word.len();
+        let entire_content_matched = prefix_length == self.content.len();
+
+        if entire_word_matched {
+            if entire_content_matched {
+                return (self.is_present, matched, &self);
+            }
+
+            return (false, matched, &self);
+        }
+
+        if !entire_content_matched {
+            return (false, matched, &self);
+        }
+
+        let next_node = self.child_nodes.get(&word[prefix_length]);
+
+        match next_node {
+            None => (false, matched, &self),
+            Some(next_node) => next_node.get_max_match(&word[prefix_length..]),
+        }
+    }
+
+    pub fn descendant_suffixes<'a>(&'a self) -> Box<dyn Iterator<Item = Bytes> + 'a> {
+        let child_nodes = self.child_nodes.values();
+
+        let suffixes = std::iter::once(self.content.clone())
+            .filter(|_x| self.is_present)
+            .chain(child_nodes.flat_map(|node| {
+                node.descendant_suffixes()
+                    .map(|suffix| [self.content.clone(), suffix].concat().into())
+            }));
+
+        Box::new(suffixes)
+    }
+}
+
+#[derive(Debug)]
+pub struct RadixTree {
+    root: RadixNode,
+}
+
+impl RadixTree {
+    pub fn root(&self) -> &RadixNode {
+        &self.root
+    }
+
+    pub fn add_word(&mut self, word: Bytes) {
+        self.root.add_word(word);
+    }
+}
+
+impl Default for RadixTree {
+    fn default() -> Self {
+        RadixTree {
+            root: RadixNode::new(Bytes::new(), false, 0),
+        }
+    }
+}
+
+fn get_shared_prefix_length(a: &[u8], b: &[u8]) -> usize {
+    let max_index = std::cmp::min(a.len(), b.len());
```

Comment:
```rust
a.iter().zip(b.iter()).enumerate().find(|(_, (a, b))| a != b).map_or_else(|| a.len().min(b.len()), |(idx, _)| idx)
```

`zip` limits to the shortest length

Reply:
Thanks, done in 01080aaca06e499c7f1dff15a4d64aedfa01ac2b